dev/PushNotifications/PushNotificationsLongRunningTask/platform.cpp

```diff
@@ -49,24 +60,115 @@ STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterF
 {
     auto lock = m_lock.lock_shared();
     RETURN_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
-    
+
+    // To be completed after Sharath's PR
+
     return E_NOTIMPL;
 }
 
-STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterForegroundActivator(_In_ IWpnForegroundSink* sink, _In_ PCWSTR processName)
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterActivator(_In_ PCWSTR processName) noexcept try
+{
+    auto lock = m_lock.lock_shared();
+    RETURN_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    std::wstring appId = GetAppIdentifier(processName);
+    m_notificationListenerManager.AddListener(appId, processName);
+
+    if (!m_notificationListenerManager.IsEmpty())
+    {
+        m_lifetimeManager.Cancel();
+    }
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::UnregisterActivator(_In_ PCWSTR processName) noexcept try
+{
+    auto lock = m_lock.lock_shared();
+    RETURN_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    m_notificationListenerManager.RemoveListener(processName);
+    m_foregroundSinkManager->Remove(processName);
+
+    // TODO: Remove the app from storage
+
+    if (m_notificationListenerManager.IsEmpty())
+    {
+        m_lifetimeManager.Setup();
+    }
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterForegroundActivator(_In_ IWpnForegroundSink* sink, _In_ PCWSTR processName) noexcept try
 {
     auto lock = m_lock.lock_exclusive();
     RETURN_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
 
-    m_foregroundSinkManager.Add(processName, sink);
+    std::wstring appId = GetAppIdentifier(processName);
+
+    m_notificationListenerManager.AddListener(appId, processName);
+    m_foregroundSinkManager->Add(processName, sink);
     return S_OK;
 }
+CATCH_RETURN()
 
-STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::UnregisterForegroundActivator(_In_ PCWSTR processName)
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::UnregisterForegroundActivator(_In_ PCWSTR processName) noexcept try
 {
     auto lock = m_lock.lock_exclusive();
     RETURN_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
 
-    m_foregroundSinkManager.Remove(processName);
+    m_foregroundSinkManager->Remove(processName);
     return S_OK;
 }
+CATCH_RETURN()
+
+// Returns a map of key-value pairs, where key is appId and value is processName.
+std::map<std::wstring, std::wstring> NotificationsLongRunningPlatformImpl::GetFullTrustApps()
+{
+    std::map<std::wstring, std::wstring> mapOfFullTrustApps;
+
+    // Get list of full trust apps with valid channels from wpncore
+    wil::unique_cotaskmem_array_ptr<PWSTR> appIds;
+    PushNotifications_GetFullTrustApplicationsWithChannels(appIds.addressof(), appIds.size_address<ULONG>());
+
+    // Load list of apps from Storage
+    auto values{ m_storage.Values() };
```

Comment:
No locks? If you put a lock here, please make sure that the caller does not share the same lock(deadlock case)

Reply:
It should only called during Initialize(), which already acquired a lock. I added a comment explaining this.